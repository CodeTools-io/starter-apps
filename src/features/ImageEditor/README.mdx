---
type: component
id: ImageEditor
name: Image Editor
categoryId: feature
icon: Image
description: Description still needed
path: /image-editor
domain:
  - name: EntityName
  - description: A description of the entity
components:
  ImageEditor: The default export and wraps all other components.
  ImageEditorHeader: Container for actions that can be performed on the image.
  ImageEditorFooter: Container for metadata about the image.
  ImageEditorCanvas: Container for the image being edited.
  ImageEditorMain: Container for the the canvas, header, and footer.
  ImageEditorSidebar: Container for settings and actions that can be performed on the image.
grommet:
  - Box
  - Button
  - Heading
  - Layer
  - Paragraph
---

# Image Editor

Description still needed
