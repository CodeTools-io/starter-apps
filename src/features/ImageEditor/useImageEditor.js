import { useMemo, useState } from 'react';
import * as config from './config';

export default function useImageEditor() {
  const [layers, setLayers] = useState(config?.layers);

  return useMemo(() => {
    return { layers };
  }, [layers]);
}
