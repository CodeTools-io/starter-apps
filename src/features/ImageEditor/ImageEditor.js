import React from 'react';
import { Box, Grid, Heading } from 'grommet';
import useImageEditor from './useImageEditor';

function ImageEditorHeader({ children, ...props }) {
  return <Box {...props}>{children}</Box>;
}

function ImageEditorFooter({ children, ...props }) {
  return <Box {...props}>{children}</Box>;
}

function ImageEditorCanvas({ children, ...props }) {
  return (
    <Box
      height={{ maxHeight: '100%' }}
      overflow="auto"
      border={{ size: 'large', color: 'blue' }}
      fill
      {...props}
    >
      {children}
    </Box>
  );
}

function ImageEditorMain({ layers, ...props }) {
  return (
    <Box {...props}>
      <ImageEditorHeader gridArea="header">header</ImageEditorHeader>
      <ImageEditorCanvas gridArea="canvas">
        {layers?.map((layer) => {
          return <ImageEditorLayer key={layer?.id} {...layer} />;
        })}
      </ImageEditorCanvas>
      <ImageEditorFooter gridArea="footer">footer</ImageEditorFooter>
    </Box>
  );
}

function ImageEditorSidebar({ layers, ...props }) {
  return (
    <Box {...props}>
      <Heading level={3}>Sidebar</Heading>
    </Box>
  );
}

function ImageEditorLayer({ name, type, settings = {} }) {
  switch (type) {
    case 'image':
      return (
        <Box border={{ size: 'large', color: 'red' }}>
          <img alt={name} {...settings} />
        </Box>
      );

    default:
      return <Box>{name}</Box>;
  }
}
export default function ImageEditor({ ...props }) {
  const { layers } = useImageEditor();

  return (
    <Box
      className="ImageEditor"
      direction="row"
      overflow="hidden"
      border={{ size: 'large', color: 'orange' }}
      fill
      {...props}
    >
      <ImageEditorMain layers={layers} flex />

      <ImageEditorSidebar layers={layers} width="medium" />
    </Box>
  );
}
