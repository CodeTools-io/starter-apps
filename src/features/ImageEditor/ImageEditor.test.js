import React from 'react';
import ImageEditor from './ImageEditor';

describe('ImageEditor', () => {
  test('can render', () => {
    const { container } = renderContainer(<ImageEditor />);

    expect(container).toMatchSnapshot();
  });
});
