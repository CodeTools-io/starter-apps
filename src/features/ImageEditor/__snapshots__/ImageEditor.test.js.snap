// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageEditor can render 1`] = `
<div>
  <div
    class="ErrorBoundary"
  >
    <div
      class="StyledGrommet-sc-19lkkz7-0 djOnwZ"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 dYebPD ImageEditor"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 dYebPD ImageEditorMain"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 giJDfc ImageEditorCanvas"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 bxevCE"
            >
              <canvas
                style="min-width: 800px; max-width: 800px; min-height: 533px; max-height: 533px;"
              >
                Could not load canvas
              </canvas>
            </div>
          </div>
          <div
            class="StyledBox-sc-13pk1d4-0 dYebPD ImageEditorFooter"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 dYebPD"
            />
            <div
              class="StyledBox-sc-13pk1d4-0 gaMuRK"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 cUsvtw"
              >
                <div
                  class="StyledBox-sc-13pk1d4-0 bNedl"
                >
                  <button
                    aria-label="Open Drop"
                    class="StyledButton-sc-323bzc-0 eIqAvo"
                    name="blur"
                    type="button"
                  >
                    Blur
                  </button>
                  <div
                    class="StyledBox__StyledBoxGap-sc-13pk1d4-1 bUcUJh"
                  />
                  <button
                    aria-label="Open Drop"
                    class="StyledButton-sc-323bzc-0 eIqAvo"
                    name="brightness"
                    type="button"
                  >
                    Brightness
                  </button>
                  <div
                    class="StyledBox__StyledBoxGap-sc-13pk1d4-1 bUcUJh"
                  />
                  <button
                    aria-label="Open Drop"
                    class="StyledButton-sc-323bzc-0 eIqAvo"
                    name="contrast"
                    type="button"
                  >
                    Contrast
                  </button>
                  <div
                    class="StyledBox__StyledBoxGap-sc-13pk1d4-1 bUcUJh"
                  />
                  <button
                    aria-label="Open Drop"
                    class="StyledButton-sc-323bzc-0 eIqAvo"
                    name="grayscale"
                    type="button"
                  >
                    Grayscale
                  </button>
                  <div
                    class="StyledBox__StyledBoxGap-sc-13pk1d4-1 bUcUJh"
                  />
                  <button
                    aria-label="Open Drop"
                    class="StyledButton-sc-323bzc-0 eIqAvo"
                    name="hue"
                    type="button"
                  >
                    Hue
                  </button>
                  <div
                    class="StyledBox__StyledBoxGap-sc-13pk1d4-1 bUcUJh"
                  />
                  <button
                    aria-label="Open Drop"
                    class="StyledButton-sc-323bzc-0 eIqAvo"
                    name="invert"
                    type="button"
                  >
                    Invert
                  </button>
                  <div
                    class="StyledBox__StyledBoxGap-sc-13pk1d4-1 bUcUJh"
                  />
                  <button
                    aria-label="Open Drop"
                    class="StyledButton-sc-323bzc-0 eIqAvo"
                    name="opacity"
                    type="button"
                  >
                    Opacity
                  </button>
                  <div
                    class="StyledBox__StyledBoxGap-sc-13pk1d4-1 bUcUJh"
                  />
                  <button
                    aria-label="Open Drop"
                    class="StyledButton-sc-323bzc-0 eIqAvo"
                    name="saturate"
                    type="button"
                  >
                    Saturate
                  </button>
                  <div
                    class="StyledBox__StyledBoxGap-sc-13pk1d4-1 bUcUJh"
                  />
                  <button
                    aria-label="Open Drop"
                    class="StyledButton-sc-323bzc-0 eIqAvo"
                    name="sepia"
                    type="button"
                  >
                    Sepia
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
